m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Phaser;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An M3 implementation of a {@link CachedStatsReporter}.
+ */
+public class M3Reporter implements CachedStatsReporter, AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(M3Reporter.class);
+
+    private static final int DEFAULT_METRIC_SIZE = 100;
+
+    private static final int DEFAULT_MAX_QUEUE_SIZE = 4096;
+    private static final int DEFAULT_MAX_PACKET_SIZE = 1440;
+
+    private static final int THREAD_POOL_SIZE = 10;
+
+    private static final String SERVICE_TAG = "service";
+    private static final String ENV_TAG = "env";
+    private static final String HOST_TAG = "host";
+    private static final String DEFAULT_TAG_VALUE = "default";
+
+    private static final String DEFAULT_HISTOGRAM_BUCKET_ID_NAME = "bucketid";
+    private static final String DEFAULT_HISTOGRAM_BUCKET_NAME = "bucket";
+    private static final int DEFAULT_HISTOGRAM_BUCKET_TAG_PRECISION = 6;
+
+    private static final int EMIT_METRIC_BATCH_OVERHEAD = 19;
+    private static final int MIN_METRIC_BUCKET_ID_TAG_LENGTH = 4;
+
+    private M3.Client client;
+
+    private final MetricBatch metricBatch;
+
+    private final Object calcLock = new Object();
+    private TCalcTransport calc;
+    private TProtocol calcProtocol;
+
+    private int freeBytes;
+
+    private ResourcePool resourcePool;
+
+    private String bucketIdTagName;
+    private String bucketTagName;
+    private String bucketValFmt;
+
+    // Holds metrics to be flushed
+    private BlockingQueue<SizedMetric> metricQueue;
+    // The service used to execute metric flushes
+    private ExecutorService executor;
+    // Used to keep track of how many processors are in progress in
+    // order to wait for them to finish before closing
+    private Phaser phaser = new Phaser(1);
+
+    // Use inner Builder class to construct an M3Reporter
+    private M3Reporter(Builder builder) {
+        try {
+            // Builder verifies non-null, non-empty hostPorts
+            String[] hostPorts = builder.hostPorts;
+
+            TProtocolFactory protocolFactory;
+            if (builder.protocol == ThriftProtocol.COMPACT) {
+                protocolFactory = new TCompactProtocol.Factory();
+            } else {
+                protocolFactory = new TBinaryProtocol.Factory();
+            }
+
+            TTransport transport = new TUdpClient(hostPorts[0]);
+
+            client = new M3.Client(protocolFactory.getProtocol(transport));
+            resourcePool = new ResourcePool(builder.maxQueueSize, protocolFactory);
+
+            Set<MetricTag> commonTags = toMetricTagsToSet(builder.commonTags);
+
+            // Set and ensure required tags
+            if (!builder.commonTags.containsKey(SERVICE_TAG)) {
```

Comment:
This method is pretty long already, instead of checking these in the ctor, could you move these to the builder?